File: javascript/markdown/markdown.js
function wrap(text, tag) {
    return `<${tag}>${text}</${tag}>`;
}

function parser(markdown, delimiter, tag) {
    const pattern = new RegExp(`${delimiter}(.+)${delimiter}`);
    const replacement = `<${tag}>$1</${tag}>`;
    return markdown.replace(pattern, replacement);
}

function parseText(markdown) {
    const parsedStrong = parser(markdown, '__', 'strong');
    return parser(parsedStrong, '_', 'em');
}

function parseHeader(markdown) {
    let count = /^\#+/.exec(markdown)[0].length;
    const headerTag = `h${count}`;
    return wrap(markdown.substring(count + 1), headerTag);
}

function parseListItem(markdown) {
    return wrap(parseText(markdown.substring(2)), 'li');
}

function parseParagraph(markdown) {
    return wrap(parseText(markdown), 'p');
}

function parseLine(markdown) {

    if (/^\#{1,6} /.test(markdown)) {
        return parseHeader(markdown);
    }

    if (markdown.startsWith('*')) {
        return parseListItem(markdown);
    } 

    return parseParagraph(markdown);
}

export function parse(markdown) {
    return markdown.split('\n')
            .map(parseLine).join('')
            .replace(new RegExp('(<li>.+</li>)'), '<ul>$1</ul>');
}

File: python/markdown/markdown.py
import re


def parse_header(markdown: str) -> str:
    m = re.match(r'^(#{1,6})\s(.*)$', markdown)
    return f'<h{len(m.group(1))}>{m.group(2)}</h{len(m.group(1))}>'


def parse_bold_and_italics(markdown: str) -> str:

    markdown = re.sub(r'__(.*)__', r'<strong>\1</strong>', markdown)
    markdown = re.sub(r'_(.*)_', r'<em>\1</em>', markdown)

    return markdown


def parse_line(markdown: str) -> str:
    if re.match(r'^#{1,6}\s.*$', markdown):
        return parse_header(markdown)

    markdown = parse_bold_and_italics(markdown)

    if re.match(r'\* .*', markdown):
        return f'<ul><li>{markdown[2:]}</li></ul>'

    return f'<p>{markdown}</p>'


def parse(markdown):
    res = ''.join([parse_line(line) for line in markdown.split('\n')])
    return res.replace('</ul><ul>', '')
